### DISCUSSION TITLE
Lesson 7: Passing priceFeed as parameter in AggregatorV3Interface

### DISCUSSION THREAD

**Discussion by codeplay22** [Cyfrin/foundry-full-course-cu#1484] [Q&A] [answered]
This is my FundMe.sol

I wanted to know how **AggregatorV3Interface(priceFeed);** is accepting priceFeed in it as a parameter when **AggregatorV3Interface.sol** doesn't have any constructors in it or accepts any parameters in the Interface.How is the code working?

```  // SPDX-License-Identifier: MIT
// 1. Pragma
pragma solidity ^0.8.19;
// 2. Imports

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

// 3. Interfaces, Libraries, Contracts
error FundMe__NotOwner();

/**
 * @title A sample Funding Contract
 * @author Patrick Collins
 * @notice This contract is for creating a sample funding contract
 * @dev This implements price feeds as our library
 */
contract FundMe {
    // Type Declarations
    using PriceConverter for uint256;

    // State variables
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    address private immutable i_owner;
    address[] private s_funders;
    mapping(address => uint256) private s_addressToAmountFunded;
    AggregatorV3Interface private s_priceFeed;

    // Events (we have none!)

    // Modifiers
    modifier onlyOwner() {
        // require(msg.sender == i_owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }

    // Functions Order:
    //// constructor
    //// receive
    //// fallback
    //// external
    //// public
    //// internal
    //// private
    //// view / pure

    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;
    }

    /// @notice Funds our contract based on the ETH/USD price
    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < s_funders.length; funderIndex++) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // Transfer vs call vs Send
        // payable(msg.sender).transfer(address(this).balance);
        (bool success,) = i_owner.call{value: address(this).balance}("");
        require(success);
    }

    function cheaperWithdraw() public onlyOwner {
        address[] memory funders = s_funders;
        // mappings can't be in memory, sorry!
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // payable(msg.sender).transfer(address(this).balance);
        (bool success,) = i_owner.call{value: address(this).balance}("");
        require(success);
    }

    /** Getter Functions */

    /**
     * @notice Gets the amount that an address has funded
     *  @param fundingAddress the address of the funder
     *  @return the amount funded
     */
    function getAddressToAmountFunded(address fundingAddress) public view returns (uint256) {
        return s_addressToAmountFunded[fundingAddress];
    }

    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }

    function getFunder(uint256 index) public view returns (address) {
        return s_funders[index];
    }

    function getOwner() public view returns (address) {
        return i_owner;
    }

    function getPriceFeed() public view returns (AggregatorV3Interface) {
        return s_priceFeed;
    }
}
```

This is the code of AggregatorV3Interface.sol

``` // SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

interface AggregatorV3Interface {
  function decimals() external view returns (uint8);

  function description() external view returns (string memory);

  function version() external view returns (uint256);

  function getRoundData(uint80 _roundId)
    external
    view
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    );

  function latestRoundData()
    external
    view
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    );
}
```

**Comment by Gloryjaw** [comment]
AggregatorV3Interface is an interface. What you are saying about constructors is true for contracts. Contracts needs to have a constructor to pass argument in them when you deploy them but same cannot be said for interfaces.

**Comment by EngrPips** [comment, marked as answer]
Hello @codeplay22 , I understand clearly what caused your confusion so let me clarify it. 

Interface is a type that let us know what are the available functions on a contract type and how we can call them, so we don't deploy interface at least as far as i know , we only use them to get the available functions and public variable from  an already deployed contract that is of the same type of the interface

so let me explain what the below code does 

```javascript

AggregatorV3Interface private s_priceFeed;

constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;
    }
```
Firstly we defined a variable named `s_priceFeed` and we stated clearly that the variable is a type of `AggregatorV3Interface` , so in the constructor we passed in an address of already deployed contract which is also a type of `AggregatorV3Interface`. on receiving the passed in address in the constructor we used it to create a contract object by wrapping the address with the `AggregatorV3Interface` interface. 

So technically by doing this   `AggregatorV3Interface(priceFeed)` we are wrapping the contract address `priceFeed` with an interface of `AggregatorV3Interface` to form a contract object  `s_priceFeed` . we do this so that we can always make call to the address using all the functions made available by the interface. I really hope i did well to make you understand. If anything still seem confusing please let me know so i explain better.

**Reply by codeplay22** [reply]
Thanks a lot for replying @EngrPips 
But then how do we get to know that if we can pass arguments in a particular Interface or not, or can we pass arguments in all Interfaces like that? 
example: randomInterface   randomVariableName;
and then i can pass that randomVariableName inside randomInterface like randomInterface(randomVariable) 
Is this true as well?

**Reply by EngrPips** [reply]
@codeplay22  we are not passing an argument to an interface, there is nothing like passing an argument to an interface, we are only wrapping a contract address with it ABI to form a contract object. You need to wrap a contract address with an ABI to form a contract object that you can call functions on because you can't call function on an address.

In your explanation above `randomInterface` would be the interface of the `randomVariable` contract address, we can't just use only `randomVariable` contract address to make call as it is only an address on the where a contract is deployed , so when we have that address where a contract is deployed we then wrap it with an interface so that we can direct call to that address using the functions available on the interface. You can go take a look at any interface and you would see it only contains function declarations and no implementation

**Reply by codeplay22** [reply]
Hi @EngrPips i hope i'm not frustating you by asking too many questions on one point haha. 
i have a doubt in your 1st explaination in which you said this

```Firstly we defined a variable named s_priceFeed and we stated clearly that the variable is a type of AggregatorV3Interface , so in the constructor we passed in an address of already deployed contract which is also a type of AggregatorV3Interface. on receiving the passed in address in the constructor we used it to create a contract object by wrapping the address with the AggregatorV3Interface interface.```


So my question is that when you said  **in constructor we passed in an address of already deployed contract which is also a type of AggregatorV3Interface**  how ```priceFeed``` is a type of ```AggregatorV3Interface? ```

This is my DeployFundMe.s.sol

```// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.19;

import {Script} from "forge-std/Script.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {FundMe} from "../src/FundMe.sol";

contract DeployFundMe is Script {
    function run() external returns (FundMe, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig(); // This comes with our mocks!
        address priceFeed = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        FundMe fundMe = new FundMe(priceFeed);
        vm.stopBroadcast();
        return (fundMe, helperConfig);
    }
}
```

And this is my HelperConfig.s.sol

```// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {MockV3Aggregator} from "../test/mock/MockV3Aggregator.sol";
import {Script} from "forge-std/Script.sol";

contract HelperConfig is Script {
    NetworkConfig public activeNetworkConfig;

    uint8 public constant DECIMALS = 8;
    int256 public constant INITIAL_PRICE = 2000e8;

    struct NetworkConfig {
        address priceFeed;
    }

    event HelperConfig__CreatedMockPriceFeed(address priceFeed);

    constructor() {
        if (block.chainid == 11155111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else {
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory sepoliaNetworkConfig) {
        sepoliaNetworkConfig = NetworkConfig({
            priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306 // ETH / USD
        });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory anvilNetworkConfig) {
        // Check to see if we set an active network config
        if (activeNetworkConfig.priceFeed != address(0)) {
            return activeNetworkConfig;
        }
        vm.startBroadcast();
        MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
            DECIMALS,
            INITIAL_PRICE
        );
        vm.stopBroadcast();
        emit HelperConfig__CreatedMockPriceFeed(address(mockPriceFeed));

        anvilNetworkConfig = NetworkConfig({priceFeed: address(mockPriceFeed)});
    
    }
}
```
isn't priceFeed is a type of  ``` MockV3Aggregator```  instead of  ```AggregatorV3Interface``` ?

**Reply by EngrPips** [reply]
Hello @codeplay22  `MockV3Aggregator` and `AggregatorV3Interface` are similar, it's just that we are using `MockV3Aggregator` on local anvil chain and `AggregatorV3Interface` on real test net. The two interface have almost the same set of functions declarations. You should open them up and see for yourself.  `MockV3Aggregator` is the mock version of `AggregatorV3Interface`. 

And no you are not frustrating me but making me learn more.

**Reply by codeplay22** [reply]
@EngrPips Thanks a lot for clarifying my doubts brother.
